### ISSUE TITLE
Feature request: return a vector of keys instead of a data.table when conditioning in j 

### ISSUE THREAD

**Issue by MichaelChirico** [Rdatatable/data.table#1242]
This is really minor, but it's been bothering me a bit lately as I've been using the workaround a lot.

It would be nice to have a simple way to tell `data.table` to return only a vector of the `by` keys satisfying a condition in `j` instead of returning a `data.table` with two identical columns and extracting.

For example, take this `data.table`:

    dt<-data.table(id=rep(1:3,1:3),var=runif(6),key="id")

I want the `id`s for which I have at least two observations. It would be nice if this worked:

    dt[,if (.N>=2) id,by=id]

But we get a `data.table` instead of a vector:

       id V1
    1:  2  2
    2:  3  3

What makes this strange is that we're doubling the information; this is solved pretty easily by simply extracting with `$id` or `$V1` (the latter usually being shorter for real-world ID variables).

However, I might have expected one of these work:

    dt[,if (.N>=2),by=id] #error, need to supply a result to the if condition
    dt[,if(.N>=2)NULL,by=id] #empty data.table

Or perhaps even:

    dt[,if(.N<2)NULL,by=id]    #empty data.table

Not sure if there's a simple way to allow for syntax like the above to result in the desired output _without_ overcomplicating things, but it would be nice to see a simple way to do this.

**Comment by franknarf1** [comment]
I would use a feature like this, but could also do without it: `dt[,if(.N>=2)1,by=id]$id`

I'm not crazy about the implementation suggested, since I don't want to see special parsing of `j` beyond treatment of `:=` and `.()`. It's complicated enough having `i` parsed specially to guess whether a join is desired. 

I'd suggest instead that the `having` arg be added https://github.com/Rdatatable/data.table/issues/788 and that `j` be allowed to be missing when `having` is present, so that this works:

    dt[by=id,having=(.N>=2)]
    # [1] 2 3

**Comment by eantonya** [comment]
I think `dt[, NULL, by = id]` returning a `data.table` with just the `id` column seems fairly reasonable. I don't see anything immediately wrong/inconsistent with that.

**Comment by eantonya** [comment]
Nm, I see now what the problem would be - expressions like `dt[, if (.N > 2) .SD, by = id]` rely on `dt[, NULL, by = id]` returning 0 rows, so they would break.

Basically the problem is the following:

```R
fn = function(a) if (a > 2) NULL

is.null(fn(1))
#[1] TRUE

is.null(fn(4))
#[1] TRUE
```

There is one thing that's different though between `fn(1)` and `fn(4)` - one of them returns `NULL` invisibly, but I don't know how to check that off the top of my head.

**Comment by gsee** [comment]
Depending on how it's implemented, it's likely that `dt[, NULL, by = id]` returning anything other than `NULL` would break some of my code.  I expect that to return NULL, so if I might later check the length and only do further processing if `is.null` is TRUE (or the `length` or `NROW` is greater than zero).  Maybe if that only happens when `NULL` is passed into `j` it wouldn't be problematic, but if there are nested `[.data.table` calls, it seems like it could be a problem.

**Comment by franknarf1** [comment]
An alternative would be to have `dt[,.BY,by=id]` indicate that only the `by` column is desired. As far as the "having" thing goes, I'd still opt for a separate argument `dt[,.BY,by=id,having=(.N>2)]`, though I guess `dt[,if(.N>2).BY,by=id]` wouldn't be bad either (as long as it's not too much work to implement and maintain such special parsing of `j`).

No one ever uses `.BY`, and certainly not on its own, so I wouldn't expect this to break much code.

**Comment by eantonya** [comment]
@franknarf1 I don't see how any of those expressions can be understood by a regular human to give the desired result ;)

**Comment by arunsrinivasan** [comment]
Not everything needs to be within one `[]`.

**Comment by franknarf1** [comment]
@eantonya A fair point. :) Just brainstorming. It is true that overloading the `j=NULL` behavior wouldn't work... just trying to think of what could.

@arunsrinivasan Yeah, this may be stretching it beyond where it needs to go.

**Comment by MichaelChirico** [comment]
@arunsrinivasan yes, but typically when I'm using this it's of the form `dt[.(dt[,if(.N>=2)id,by=id]$id)]`

**Comment by eantonya** [comment]
@MichaelChirico why are you doing that instead of `dt[, if(.N >= 2) .SD, by = id]`?

**Comment by MichaelChirico** [comment]
@eantonya sorry, should have been more clear, edited.

**Comment by arunsrinivasan** [comment]
@MichaelChirico you don't need to subset at all there.

```r
dt[dt[, if (.N >= 2L) id, by=id], var := ...]
```

would join on `id` and work just fine. But why are you joining here at all?

```r
dt[, var := if (.N >= 2L) ... , by=id]
```
What's wrong with this? It's much more straightforward as to what you're trying to do.

Two additional points:

1. I think this is very closely related to your other post on only returning *strictly* unique rows (except this is the opposite) - https://github.com/Rdatatable/data.table/issues/1163. 

`group_by` returning groups is the most common scenario, and having an argument just for that to return a vector seems unnecessary to me.

2. Please use spaces in your code. It's extremely hard to parse/understand.

**Comment by eantonya** [comment]
Closing per the above discussion (at least until a better suggestion comes along).
